### ISSUE TITLE
RSS usage spikes w/ multiGet usage

### ISSUE THREAD

**Issue by WPprodigy** [memcached/memcached#1046]
Firstly, just wanted to say thanks for maintaining this project :)

## Describe the bug
Similar to https://github.com/memcached/memcached/issues/895, we noticed that RSS usage increases quite a bit after enabling multiGet support on the client side. Increased to the point where the memcached pods were exceeding their memory allocations (RAM allocated for item storage + a bit of padding) and the processes were being killed off by the OOM Killer.

![RSS](https://github.com/memcached/memcached/assets/8536129/07527aef-3e22-467f-aadc-bcb008865c72)
(left graph shows `memcached_current_bytes` & `memcached_limit_bytes` as gathered from memcached stats, right graph is the actual RSS usage on the server)

## To Reproduce

We've been experiencing this across a large number of environments/sites on our infrastructure, and after reading through the code a bit and doing some testing I believe it comes down to these three main factors:

- We use persistent connections client-side ([PHP client](https://www.php.net/manual/en/book.memcached.php)), so each connection will handle a few requests before the php-fpm child worker is recycled. Notably, this means we can't `$mc->close()` the connections.
- There is a high max_connections limit (32768).
- There are "large" ascii multigets (the only thing that changed)

My current best guess is that the [separate code path](https://github.com/memcached/memcached/blob/e0e415b7b2b43a6ddd01a9c3ad45fb46358d526b/proto_text.c#L472-L481) hits for "large ascii gets" and memory is allocated differently for the response object, and appears to be stored per request rather than per thread. Since we use persistent connections, `conn_close()` is called less often (not at all?) and thus `rbuf_release()` [doesn't really get called](https://github.com/memcached/memcached/blob/e0e415b7b2b43a6ddd01a9c3ad45fb46358d526b/memcached.c#L920) for the extra allocated memory on the request and it ends up being claimed for a long time. And then I assume there's some connection reclaiming logic that eventually does close out connections once the max starts getting hit. But up until `curr_connections` reaches  `max_connections`, the `read_buf_bytes` is just ever-increasing. Then it hits it's max and stays static.

So as a result, you more or less end up with extra persistent memory usage on the server that is essentially: `(size of largest multiGet) * (max_connections)`. Which doesn't feel like a super great usage of memory 😬.

At first I thought `read_buf_mem_limit` would be a good solution, but after reading through some of the code I'm not so sure anymore as it seems to handle the response buffers that are on the thread, but not the "malloc'ed ascii multiGet per-request" buffers? Let me know if I'm wrong though!

## Details

Here's are some related details from a server we did some testing on:

<details>
  <summary>Before changing max_connections</summary>
  
With max_connections set to 32768, and after waiting a while to exceed that number (total_connections), the `read_buf_bytes` will hit `1649344512` and remain static from my observations.
  
  ```php
["uptime"] => int(58527)
["time"] => int(1686346250)
["version"] => string(6) "1.6.17"
["libevent"] => string(13) "2.1.12-stable"
["pointer_size"] => int(64)
["rusage_user"] => float(3815.643834)
["rusage_system"] => float(10106.782624)

["max_connections"] => int(32768)
["curr_connections"] => int(228)
["total_connections"] => int(46531)
["rejected_connections"] => int(0)
["connection_structures"] => int(520)

["response_obj_oom"] => int(0)
["response_obj_count"] => int(1)
["response_obj_bytes"] => int(98304)
["read_buf_count"] => int(100668)
["read_buf_bytes"] => int(1649344512)
["read_buf_bytes_free"] => int(1649229824)
["read_buf_oom"] => int(0)
["threads"] => int(6)

["bytes"] => int(4206360440)
["curr_items"] => int(7848134)
["total_items"] => int(14157074)

["cmd_get"] => int(298502468)
["cmd_set"] => int(21484210)
["bytes_read"] => int(31594629174)
["bytes_written"] => int(344009993641)
["limit_maxbytes"] => int(6442450944)
  ```
</details>

<details>
  <summary>After changing max_connections</summary>
  
After halfing max_connections to 16384, and after waiting a while to exceed that number again (total_connections), the `read_buf_bytes` hits around half of the previous at `788496384` and again remains static from my observations. There's of course a bit of a time gap in these reporting so client-side data is changing but the results remain the same in that `read_buf_bytes` does hit a "max" and then remains there until the server crashes / is rebooted. And the amount of read buffer bytes definitely goes down alongside `max_connections`.
  
  ```php
	["uptime"] => int(54060)
	["rusage_user"] => float(3695.635085)
	["rusage_system"] => float(10295.950047)

	["max_connections"] => int(16384)
	["curr_connections"] => int(247)
	["total_connections"] => int(40127)
	["rejected_connections"] => int(0)
	["connection_structures"] => int(367)

	["response_obj_oom"] => int(0)
	["response_obj_count"] => int(1)
	["response_obj_bytes"] => int(98304)
	["read_buf_count"] => int(48126)
	["read_buf_bytes"] => int(788496384)
	["read_buf_bytes_free"] => int(788381696)
	["read_buf_oom"] => int(0)
	["threads"] => int(6)

	["bytes"] => int(3986392337)
	["curr_items"] => int(6748437)
	["total_items"] => int(12232185)

	["cmd_get"] => int(297327263)
	["cmd_set"] => int(18400080)
	["bytes_read"] => int(59914141483)
	["bytes_written"] => int(765239705287)
	["limit_maxbytes"] => int(6442450944)
  ```
</details>

**Comment by dormando** [comment]
https://github.com/memcached/memcached/blob/e0e415b7b2b43a6ddd01a9c3ad45fb46358d526b/memcached.c#L3099 this is the normal `rbuf_release()` call, when the client socket goes idle, but for multigets that's not where the memory all goes.

rbuf allocations are internally used for `mc_resp` response objects, which are ~1k objects that describe the response for each key being fetched. If you multiget 1000 keys it will use 1000 kilobytes of response memory. Since the ascii multiget command cannot begin to return data until it has read all keys from the client, it will buffer memory forever. It does not however suffer from the TCP deadlocking issue I described in my prior comment but the tradeoff is OOM so I've never liked it much.

In the newer meta protocol this syntax doesn't exist at all, and you are required to pipeline simple commands for batch fetches.

**Comment by dormando** [comment]
To put this another way:

- rbuf memory is only allocated from the OS when the existing pool is 100% used.
- so if you're allocating enough memory to cause an OOM, you would still OOM.
- we can't arbitrarily set an rbuf memory limit for users by default, for various reasons.

I can still audit for bugs, but if you are graphing or storing that information, you should see the "rbuf_bytes_free" metrics hit 0 every time the "rbuf_bytes" metric increases"

**Comment by WPprodigy** [comment]
Thanks for the replies!

I think that makes sense. So is it true to say that the response buffers even for ascii multigets are supposed to be claimed per thread (not per request)? Or in other words, the "largest multiGet" will claim X memory and remain claimed forever on that thread for the remainder of it's life. But the overall claimed memory for these buffers should never exceed the `(largest multiGet) * (# of threads)`?

What left me with thinking this might be a bug was following the code a bit and seeing `do_cache_free()` called when it's time to cleanup, and that's when `read_buf_mem_limit` would do it's cleanup if set. But the callers don't always appear to account for the `rbuf_malloced` condition like we see in `rbuf_release()` (which is only called when connections are closed or sockets go idle). So it kinda felt like the malloced response buffer might just end up sticking around much longer in this case. (Caveat: these are heavily guesses, I don't really know C 😅).
____________________

I guess it just seems suspicious that when we decrease `max_connections` we also see that the max amount for `read_buf_bytes` decreases (observed on multiple servers). I'll see about adding some graphing to keep an eye on if `rbuf_bytes_free` is hitting 0. 

It also takes quite a while before read_buf_bytes / rbuf_bytes_free hit their "maxes" after a server restart, it starts pretty small and just kinda keeps growing for a few hours (until seemingly total_connections >= max_connections) - I'll see about adding some graphing for that too. The request patterns for each memcached server is pretty static, as they serve web pages that are more-or-less always going to be requesting the same payloads.

**Comment by dormando** [comment]
It's a bit confusing because of performance requirements. Again, the caches are freed back to the pool when the connection becomes _idle_, not just when closed.  You can see in your stats that at rest there's no data in use.

- in older code, read and write buffers were _statically allocated_ per connection object.
- these could grow a bit, but if they grew "too much" (large multiget) they would realloc back down to a limit post-command
- which means 32,000 idle connections used `32000 * (sizeof(conn) + read buffer / write buffer)`
- with this change, an idle connection only holds `sizeof(conn)` memory, which saves gigabytes of RAM for people with high connection limits.
- to avoid a performance loss, each worker thread manages its own pool of read buffers.
- when allocating or freeing to these pools, there is no locking overhead and minimal CPU usage
- there are no conditions where these pools free back to the OS.

ascii multigets throw a wrench in everything.
- read buffers are statically sized at 16kb, which is more than enough memory for any kind of request (set requests with values will read data directly into the memory of the item it's about to store)
- multigets must be able to buffer all inbound data, which means we can't use this rbuf pool at all for _read buffers_
- so we switch to a direct malloc, which is then freed after the response is sent (see `conn_waiting` state)
- this temporarily malloced read buffer _is not counted in the stats counters_ (see below)
- after a multiget is fully read from the network, it begins allocating `mc_resp` objects for each key being returned by memcached (this holds the `VALUE` line, iovec, some other data)

However, again, most of your actual memory usage for dealing with multigets is probably in the resp objects, which draw from the read buffer pool for their memory. There is one object per key that's being stored.

So now we have a temporarily malloc'ed read buffer plus non-temporarily-malloc'ed response objects. Memcached will free (globally) the read buffer when it's done but not the response objects.

I'm still open to there being some sort of bug but I'm not seeing it yet. If there was a bug you would see the `read_buf_bytes_free` counter staying low and it would continually alloc more memory. I also run benchmark tests with multigets and have many users with multigets and the memory doesn't seem to leak.

**Comment by dormando** [comment]
also the calculation isn't as simple as `largest multiget * etc` because if you have multiple clients running multigets at once they'll stack, which is what I assume is killing you and why it tracks with max conns?

I'll read through the code again and see if there's anything obvious that could make this worse than it should be.

**Comment by dormando** [comment]
Sorry for the spam I'm trying to just fire off things as I think of them :)

You can double check for an `rbuf_malloced` leak:
- memory_malloced from slab stats
- + hash_bytes for the hash table
- + read_buf_bytes
.. should be within a few megabytes of the RSS size of the daemon. There's nearly no memory usage outside of those things outside of a couple buffers and thread stacks. If the RSS is off by hundreds of megabytes or gigabytes there could be a leak.

**Comment by WPprodigy** [comment]
Definitely appreciate all the thoughts here! Up till now, there was very little discussion/info on the internet about this xD

I don't think it's a full leak per-se, since it does appear that `read_buf_bytes` will hit a (high) max after some time and then happily stay there. (so we are able to adjust/account for this and I think the RSS will then stay steady - I'll report back when we have new data after making adjustments to prevent the crashes). But rather was just thinking that MC might not be cleaning up as often as it could be as the `read_buf_bytes` does feel to initially grow rather oddly (due to us not closing connections client-side was my guess).

Here are the full stats for 4 servers that service one environment in case there's something else that'd be helpful. These were taken after we dramatically increased the allowed memory (for items) to give us more time before crashes while digging into this.

<details>
  <summary>Stats</summary>
  
  ```php
array(4) {
["cache-0.redacted:11211"]= array(92) {
  ["pid"]=> int(8)
  ["uptime"]=> int(58478)
  ["time"]=> int(1686346250)
  ["version"]=> string(6) "1.6.17"
  ["libevent"]=> string(13) "2.1.12-stable"
  ["pointer_size"]=> int(64)
  ["rusage_user"]=> float(3562.71272)
  ["rusage_system"]=> float(7248.076284)
  ["max_connections"]=> int(32768)
  ["curr_connections"]=> int(228)
  ["total_connections"]=> int(46535)
  ["rejected_connections"]=> int(0)
  ["connection_structures"]=> int(539)
  ["response_obj_oom"]=> int(0)
  ["response_obj_count"]=> int(1)
  ["response_obj_bytes"]=> int(98304)
  ["read_buf_count"]=> int(63884)
  ["read_buf_bytes"]=> int(1046675456)
  ["read_buf_bytes_free"]=> int(1046560768)
  ["read_buf_oom"]=> int(0)
  ["reserved_fds"]=> int(30)
  ["cmd_get"]=> int(299411861)
  ["cmd_set"]=> int(12809027)
  ["cmd_flush"]=> int(0)
  ["cmd_touch"]=> int(0)
  ["cmd_meta"]=> int(0)
  ["get_hits"]=> int(274106021)
  ["get_misses"]=> int(25305840)
  ["get_expired"]=> int(55639)
  ["get_flushed"]=> int(0)
  ["delete_misses"]=> int(988440)
  ["delete_hits"]=> int(115089)
  ["incr_misses"]=> int(304)
  ["incr_hits"]=> int(1624701)
  ["decr_misses"]=> int(161)
  ["decr_hits"]=> int(1623371)
  ["cas_misses"]=> int(0)
  ["cas_hits"]=> int(0)
  ["cas_badval"]=> int(0)
  ["touch_hits"]=> int(0)
  ["touch_misses"]=> int(0)
  ["store_too_large"]=> int(17)
  ["store_no_memory"]=> int(0)
  ["auth_cmds"]=> int(0)
  ["auth_errors"]=> int(0)
  ["bytes_read"]=> int(32634976138)
  ["bytes_written"]=> int(391874010000)
  ["limit_maxbytes"]=> int(6442450944)
  ["accepting_conns"]=> int(1)
  ["listen_disabled_num"]=> int(0)
  ["time_in_listen_disabled_us"]=> int(0)
  ["threads"]=> int(6)
  ["conn_yields"]=> int(0)
  ["hash_power_level"]=> int(23)
  ["hash_bytes"]=> int(67108864)
  ["hash_is_expanding"]=> int(0)
  ["slab_reassign_rescues"]=> int(516)
  ["slab_reassign_chunk_rescues"]=> int(60)
  ["slab_reassign_evictions_nomem"]=> int(0)
  ["slab_reassign_inline_reclaim"]=> int(25)
  ["slab_reassign_busy_items"]=> int(0)
  ["slab_reassign_busy_deletes"]=> int(0)
  ["slab_reassign_running"]=> int(0)
  ["slabs_moved"]=> int(67)
  ["lru_crawler_running"]=> int(0)
  ["lru_crawler_starts"]=> int(51864)
  ["lru_maintainer_juggles"]=> int(22505104)
  ["malloc_fails"]=> int(0)
  ["log_worker_dropped"]=> int(0)
  ["log_worker_written"]=> int(0)
  ["log_watcher_skipped"]=> int(0)
  ["log_watcher_sent"]=> int(0)
  ["log_watchers"]=> int(0)
  ["unexpected_napi_ids"]=> int(0)
  ["round_robin_fallback"]=> int(0)
  ["bytes"]=> int(3796467697)
  ["curr_items"]=> int(7169989)
  ["total_items"]=> int(12773872)
  ["slab_global_page_pool"]=> int(0)
  ["expired_unfetched"]=> int(2727589)
  ["evicted_unfetched"]=> int(0)
  ["evicted_active"]=> int(0)
  ["evictions"]=> int(0)
  ["reclaimed"]=> int(4152)
  ["crawler_reclaimed"]=> int(3814339)
  ["crawler_items_checked"]=> int(2700633222)
  ["lrutail_reflocked"]=> int(16272815)
  ["moves_to_cold"]=> int(34440861)
  ["moves_to_warm"]=> int(29164231)
  ["moves_within_lru"]=> int(9002965)
  ["direct_reclaims"]=> int(0)
  ["lru_bumps_dropped"]=> int(0)
  }


["cache-1.redacted:11211"]= array(92) {
  ["pid"]=> int(7)
  ["uptime"]=> int(58506)
  ["time"]=> int(1686346250)
  ["version"]=> string(6) "1.6.17"
  ["libevent"]=> string(13) "2.1.12-stable"
  ["pointer_size"]=> int(64)
  ["rusage_user"]=> float(3660.78372)
  ["rusage_system"]=> float(6844.276521)
  ["max_connections"]=> int(32768)
  ["curr_connections"]=> int(227)
  ["total_connections"]=> int(46570)
  ["rejected_connections"]=> int(0)
  ["connection_structures"]=> int(548)
  ["response_obj_oom"]=> int(0)
  ["response_obj_count"]=> int(1)
  ["response_obj_bytes"]=> int(98304)
  ["read_buf_count"]=> int(99456)
  ["read_buf_bytes"]=> int(1629487104)
  ["read_buf_bytes_free"]=> int(1629372416)
  ["read_buf_oom"]=> int(0)
  ["reserved_fds"]=> int(30)
  ["cmd_get"]=> int(326624059)
  ["cmd_set"]=> int(15631497)
  ["cmd_flush"]=> int(0)
  ["cmd_touch"]=> int(0)
  ["cmd_meta"]=> int(0)
  ["get_hits"]=> int(299232731)
  ["get_misses"]=> int(27391328)
  ["get_expired"]=> int(63279)
  ["get_flushed"]=> int(0)
  ["delete_misses"]=> int(111753)
  ["delete_hits"]=> int(98812)
  ["incr_misses"]=> int(109)
  ["incr_hits"]=> int(930492)
  ["decr_misses"]=> int(0)
  ["decr_hits"]=> int(0)
  ["cas_misses"]=> int(0)
  ["cas_hits"]=> int(0)
  ["cas_badval"]=> int(0)
  ["touch_hits"]=> int(0)
  ["touch_misses"]=> int(0)
  ["store_too_large"]=> int(10)
  ["store_no_memory"]=> int(0)
  ["auth_cmds"]=> int(0)
  ["auth_errors"]=> int(0)
  ["bytes_read"]=> int(33237243690)
  ["bytes_written"]=> int(382505235112)
  ["limit_maxbytes"]=> int(6442450944)
  ["accepting_conns"]=> int(1)
  ["listen_disabled_num"]=> int(0)
  ["time_in_listen_disabled_us"]=> int(0)
  ["threads"]=> int(6)
  ["conn_yields"]=> int(0)
  ["hash_power_level"]=> int(23)
  ["hash_bytes"]=> int(67108864)
  ["hash_is_expanding"]=> int(0)
  ["slab_reassign_rescues"]=> int(26724)
  ["slab_reassign_chunk_rescues"]=> int(130)
  ["slab_reassign_evictions_nomem"]=> int(0)
  ["slab_reassign_inline_reclaim"]=> int(33)
  ["slab_reassign_busy_items"]=> int(0)
  ["slab_reassign_busy_deletes"]=> int(0)
  ["slab_reassign_running"]=> int(0)
  ["slabs_moved"]=> int(112)
  ["lru_crawler_running"]=> int(0)
  ["lru_crawler_starts"]=> int(54748)
  ["lru_maintainer_juggles"]=> int(24485136)
  ["malloc_fails"]=> int(0)
  ["log_worker_dropped"]=> int(0)
  ["log_worker_written"]=> int(0)
  ["log_watcher_skipped"]=> int(0)
  ["log_watcher_sent"]=> int(0)
  ["log_watchers"]=> int(0)
  ["unexpected_napi_ids"]=> int(0)
  ["round_robin_fallback"]=> int(0)
  ["bytes"]=> int(4250690436)
  ["curr_items"]=> int(8147225)
  ["total_items"]=> int(15600161)
  ["slab_global_page_pool"]=> int(0)
  ["expired_unfetched"]=> int(3103514)
  ["evicted_unfetched"]=> int(0)
  ["evicted_active"]=> int(0)
  ["evictions"]=> int(0)
  ["reclaimed"]=> int(4310)
  ["crawler_reclaimed"]=> int(4337634)
  ["crawler_items_checked"]=> int(3087411825)
  ["lrutail_reflocked"]=> int(22646148)
  ["moves_to_cold"]=> int(43073235)
  ["moves_to_warm"]=> int(37077171)
  ["moves_within_lru"]=> int(8197976)
  ["direct_reclaims"]=> int(0)
  ["lru_bumps_dropped"]=> int(0)
  }


["cache-2.redacted:11211"]= array(92) {
  ["pid"]=> int(7)
  ["uptime"]=> int(58543)
  ["time"]=> int(1686346249)
  ["version"]=> string(6) "1.6.17"
  ["libevent"]=> string(13) "2.1.12-stable"
  ["pointer_size"]=> int(64)
  ["rusage_user"]=> float(3466.098856)
  ["rusage_system"]=> float(7241.638033)
  ["max_connections"]=> int(32768)
  ["curr_connections"]=> int(228)
  ["total_connections"]=> int(46534)
  ["rejected_connections"]=> int(0)
  ["connection_structures"]=> int(520)
  ["response_obj_oom"]=> int(0)
  ["response_obj_count"]=> int(2)
  ["response_obj_bytes"]=> int(98304)
  ["read_buf_count"]=> int(52556)
  ["read_buf_bytes"]=> int(861077504)
  ["read_buf_bytes_free"]=> int(860946432)
  ["read_buf_oom"]=> int(0)
  ["reserved_fds"]=> int(30)
  ["cmd_get"]=> int(264494218)
  ["cmd_set"]=> int(14912492)
  ["cmd_flush"]=> int(0)
  ["cmd_touch"]=> int(0)
  ["cmd_meta"]=> int(0)
  ["get_hits"]=> int(243759870)
  ["get_misses"]=> int(20734348)
  ["get_expired"]=> int(52433)
  ["get_flushed"]=> int(0)
  ["delete_misses"]=> int(204540)
  ["delete_hits"]=> int(75185)
  ["incr_misses"]=> int(0)
  ["incr_hits"]=> int(579)
  ["decr_misses"]=> int(1)
  ["decr_hits"]=> int(0)
  ["cas_misses"]=> int(0)
  ["cas_hits"]=> int(0)
  ["cas_badval"]=> int(0)
  ["touch_hits"]=> int(0)
  ["touch_misses"]=> int(0)
  ["store_too_large"]=> int(15)
  ["store_no_memory"]=> int(0)
  ["auth_cmds"]=> int(0)
  ["auth_errors"]=> int(0)
  ["bytes_read"]=> int(26864857540)
  ["bytes_written"]=> int(281671264035)
  ["limit_maxbytes"]=> int(6442450944)
  ["accepting_conns"]=> int(1)
  ["listen_disabled_num"]=> int(0)
  ["time_in_listen_disabled_us"]=> int(0)
  ["threads"]=> int(6)
  ["conn_yields"]=> int(0)
  ["hash_power_level"]=> int(23)
  ["hash_bytes"]=> int(67108864)
  ["hash_is_expanding"]=> int(0)
  ["slab_reassign_rescues"]=> int(326)
  ["slab_reassign_chunk_rescues"]=> int(96)
  ["slab_reassign_evictions_nomem"]=> int(0)
  ["slab_reassign_inline_reclaim"]=> int(16)
  ["slab_reassign_busy_items"]=> int(0)
  ["slab_reassign_busy_deletes"]=> int(0)
  ["slab_reassign_running"]=> int(0)
  ["slabs_moved"]=> int(73)
  ["lru_crawler_running"]=> int(0)
  ["lru_crawler_starts"]=> int(53240)
  ["lru_maintainer_juggles"]=> int(22177789)
  ["malloc_fails"]=> int(0)
  ["log_worker_dropped"]=> int(0)
  ["log_worker_written"]=> int(0)
  ["log_watcher_skipped"]=> int(0)
  ["log_watcher_sent"]=> int(0)
  ["log_watchers"]=> int(0)
  ["unexpected_napi_ids"]=> int(0)
  ["round_robin_fallback"]=> int(0)
  ["bytes"]=> int(3516575183)
  ["curr_items"]=> int(6614236)
  ["total_items"]=> int(11751954)
  ["slab_global_page_pool"]=> int(0)
  ["expired_unfetched"]=> int(2529302)
  ["evicted_unfetched"]=> int(0)
  ["evicted_active"]=> int(0)
  ["evictions"]=> int(0)
  ["reclaimed"]=> int(4550)
  ["crawler_reclaimed"]=> int(3534321)
  ["crawler_items_checked"]=> int(2460294833)
  ["lrutail_reflocked"]=> int(8030448)
  ["moves_to_cold"]=> int(17955194)
  ["moves_to_warm"]=> int(13061330)
  ["moves_within_lru"]=> int(5325110)
  ["direct_reclaims"]=> int(0)
  ["lru_bumps_dropped"]=> int(0)
  }


["cache-3.redacted:11211"]= array(92) {
  ["pid"]=> int(7)
  ["uptime"]=> int(58527)
  ["time"]=> int(1686346250)
  ["version"]=> string(6) "1.6.17"
  ["libevent"]=> string(13) "2.1.12-stable"
  ["pointer_size"]=> int(64)
  ["rusage_user"]=> float(3815.643834)
  ["rusage_system"]=> float(10106.782624)
  ["max_connections"]=> int(32768)
  ["curr_connections"]=> int(228)
  ["total_connections"]=> int(46531)
  ["rejected_connections"]=> int(0)
  ["connection_structures"]=> int(520)
  ["response_obj_oom"]=> int(0)
  ["response_obj_count"]=> int(1)
  ["response_obj_bytes"]=> int(98304)
  ["read_buf_count"]=> int(100668)
  ["read_buf_bytes"]=> int(1649344512)
  ["read_buf_bytes_free"]=> int(1649229824)
  ["read_buf_oom"]=> int(0)
  ["reserved_fds"]=> int(30)
  ["cmd_get"]=> int(298502468)
  ["cmd_set"]=> int(21484210)
  ["cmd_flush"]=> int(0)
  ["cmd_touch"]=> int(0)
  ["cmd_meta"]=> int(0)
  ["get_hits"]=> int(273255831)
  ["get_misses"]=> int(25246637)
  ["get_expired"]=> int(62850)
  ["get_flushed"]=> int(0)
  ["delete_misses"]=> int(1158657)
  ["delete_hits"]=> int(102862)
  ["incr_misses"]=> int(0)
  ["incr_hits"]=> int(164906)
  ["decr_misses"]=> int(1)
  ["decr_hits"]=> int(1449)
  ["cas_misses"]=> int(0)
  ["cas_hits"]=> int(0)
  ["cas_badval"]=> int(0)
  ["touch_hits"]=> int(0)
  ["touch_misses"]=> int(0)
  ["store_too_large"]=> int(8)
  ["store_no_memory"]=> int(0)
  ["auth_cmds"]=> int(0)
  ["auth_errors"]=> int(0)
  ["bytes_read"]=> int(31594629174)
  ["bytes_written"]=> int(344009993641)
  ["limit_maxbytes"]=> int(6442450944)
  ["accepting_conns"]=> int(1)
  ["listen_disabled_num"]=> int(0)
  ["time_in_listen_disabled_us"]=> int(0)
  ["threads"]=> int(6)
  ["conn_yields"]=> int(0)
  ["hash_power_level"]=> int(23)
  ["hash_bytes"]=> int(67108864)
  ["hash_is_expanding"]=> int(0)
  ["slab_reassign_rescues"]=> int(19053)
  ["slab_reassign_chunk_rescues"]=> int(60)
  ["slab_reassign_evictions_nomem"]=> int(0)
  ["slab_reassign_inline_reclaim"]=> int(129)
  ["slab_reassign_busy_items"]=> int(0)
  ["slab_reassign_busy_deletes"]=> int(0)
  ["slab_reassign_running"]=> int(0)
  ["slabs_moved"]=> int(79)
  ["lru_crawler_running"]=> int(0)
  ["lru_crawler_starts"]=> int(52357)
  ["lru_maintainer_juggles"]=> int(27223722)
  ["malloc_fails"]=> int(0)
  ["log_worker_dropped"]=> int(0)
  ["log_worker_written"]=> int(0)
  ["log_watcher_skipped"]=> int(0)
  ["log_watcher_sent"]=> int(0)
  ["log_watchers"]=> int(0)
  ["unexpected_napi_ids"]=> int(0)
  ["round_robin_fallback"]=> int(0)
  ["bytes"]=> int(4206360440)
  ["curr_items"]=> int(7848134)
  ["total_items"]=> int(14157074)
  ["slab_global_page_pool"]=> int(0)
  ["expired_unfetched"]=> int(3006838)
  ["evicted_unfetched"]=> int(0)
  ["evicted_active"]=> int(0)
  ["evictions"]=> int(0)
  ["reclaimed"]=> int(4963)
  ["crawler_reclaimed"]=> int(4202541)
  ["crawler_items_checked"]=> int(2976372891)
  ["lrutail_reflocked"]=> int(12338764)
  ["moves_to_cold"]=> int(21197326)
  ["moves_to_warm"]=> int(15396266)
  ["moves_within_lru"]=> int(6763325)
  ["direct_reclaims"]=> int(0)
  ["lru_bumps_dropped"]=> int(0)
  }
}
  ```
</details>

**Comment by WPprodigy** [comment]
The largest multiGet on the site with those stats is also very unlikely to be more than 1mb, especially while compressed on the server. But for simple maths and to overestimate a bit, let's assume there is a 25 key multiGet that accounts for 1mb total in server-side data (and all go to the same server for simplicity). If the read buffers you're seeing make sense for that given the average active_conns, then maybe we're all good here.

**Comment by dormando** [comment]
Thanks for the stats, that is always useful. (stats settings is also nice to have if available)

I think you asked about the per worker thread pools and I didn't quite answer that well enough.

Lets say you have a multiget that comes in via a random connection and fetches 5,000 keys at once:
- will require a read buffer > 16kb so we will free ours and malloc a temporary buffer
- once read, it will allocate 5,000 `mc_resp` objects
- then it will respond to the client
- free the 5,000 `mc_resp` objects back to the _per thread pool_
- free the temporary read buffer back to the OS

This would result in the per-worker-thread cache using a bit over 5,000 kilobytes of memory after everything is done.

But, you have 6 worker threads per daemon.

So the next time the same query comes in, if it hits a different worker thread, it will allocate another 5,000 kilobytes of memory.

So your total usage will be `largest_multiget_memory_requirements * worker_thread_count`

which does suck but my hands are kind of tied, and this new memory model still uses less than it used to.

It's tough here because I'd love to share the memory pool (too slow) or free objects over some amount (has pathological use cases), or default the rbuf memory limit to 1% of item memory (pre-existing users could throw errors after upgrading).

It does seem like your multiget batch size is much larger than you're expecting though. I see ~100,000 rbuf objects allocated in one of your examples:
```
; 100000 * 15 / 6
	250000
```

That's an over-estimate but either you're doing one multiget with hundreds of thousands of keys, or there are multiple multigets with tens of thousands of keys running at the same time sometimes (big batch fetch from multiple clients?)

**Comment by WPprodigy** [comment]
> That's an over-estimate but either you're doing one multiget with hundreds of thousands of keys, or there are multiple multigets with tens of thousands of keys running at the same time sometimes (big batch fetch from multiple clients?)

I'd be quite surprised to find a multiGet with more than even 200 keys. Most definitely never more than 1000. So especially if `curr_connections * (1000 key multiGet)` isn't enough to get us to where we are, then I'm definitely suspect.

(For more context, we host thousands of enterprise WordPress applications. They all follow similar underlining DB & memcached patterns that are more or less predicable to the point where I'm quite sure there are not super massive multiGets - especially since this is fairly recently added to our platform and really only utilized from core WP in a few places wherein more than a couple hundred posts being queried at once is not really happening).

**Comment by dormando** [comment]
If you have 250 active connections and _all_ of them are doing a 1,000 key multiget simultaneously you would arrive at that correct number. However I doubt this.

Most of the times when folks have come across this issue, they had a code bug where they were accidentally fetching the same key 50,000+ times in the same query. I don't know why it's always that class of bug but it's what it is.

**Comment by WPprodigy** [comment]
So at most, the example site I've been referencing seems to spike at 306 active connections on one of the servers, and hovers around the 250:

![connections](https://github.com/memcached/memcached/assets/8536129/e2742f48-185d-4059-81c0-deed59895dc1)

And here are the stats again. You can tell by the `uptimes` how long it's been since the latest crash/restart on each. For example, cache-1 recently restarted and is at ~276mb for the `read_buf_count` right now. But that will inevitably keep growing up til a certain point.

<details>
  <summary>Fresh stats pull</summary>
 
  ```php
array( 
  ["cache-0.redacted:11211" ] = array(92) {
    ["pid"] => int(112)
    ["uptime"] => int(13218)
    ["time"] => int(1686604838)
    ["version"] => string(6) "1.6.17"
    ["libevent"] => string(13) "2.1.12-stable"
    ["pointer_size"] => int(64)
    ["rusage_user"] => float(921.488048)
    ["rusage_system"] => float(1727.995956)
    ["max_connections"] => int(16384)
    ["curr_connections"] => int(253)
    ["total_connections"] => int(10999)
    ["rejected_connections"] => int(0)
    ["connection_structures"] => int(310)
    ["response_obj_oom"] => int(0)
    ["response_obj_count"] => int(2)
    ["response_obj_bytes"] => int(98304)
    ["read_buf_count"] => int(47173)
    ["read_buf_bytes"] => int(772882432)
    ["read_buf_bytes_free"] => int(772751360)
    ["read_buf_oom"] => int(0)
    ["reserved_fds"] => int(30)
    ["cmd_get"] => int(97866790)
    ["cmd_set"] => int(3433119)
    ["cmd_flush"] => int(0)
    ["cmd_touch"] => int(0)
    ["cmd_meta"] => int(0)
    ["get_hits"] => int(81244401)
    ["get_misses"] => int(16622389)
    ["get_expired"] => int(9807)
    ["get_flushed"] => int(0)
    ["delete_misses"] => int(302401)
    ["delete_hits"] => int(34862)
    ["incr_misses"] => int(70)
    ["incr_hits"] => int(463956)
    ["decr_misses"] => int(53)
    ["decr_hits"] => int(463619)
    ["cas_misses"] => int(0)
    ["cas_hits"] => int(0)
    ["cas_badval"] => int(0)
    ["touch_hits"] => int(0)
    ["touch_misses"] => int(0)
    ["store_too_large"] => int(14)
    ["store_no_memory"] => int(0)
    ["auth_cmds"] => int(0)
    ["auth_errors"] => int(0)
    ["bytes_read"] => int(9988341023)
    ["bytes_written"] => int(98416955633)
    ["limit_maxbytes"] => int(6442450944)
    ["accepting_conns"] => int(1)
    ["listen_disabled_num"] => int(0)
    ["time_in_listen_disabled_us"] => int(0)
    ["threads"] => int(6)
    ["conn_yields"] => int(0)
    ["hash_power_level"] => int(21)
    ["hash_bytes"] => int(16777216)
    ["hash_is_expanding"] => int(0)
    ["slab_reassign_rescues"] => int(61)
    ["slab_reassign_chunk_rescues"] => int(5)
    ["slab_reassign_evictions_nomem"] => int(0)
    ["slab_reassign_inline_reclaim"] => int(2)
    ["slab_reassign_busy_items"] => int(0)
    ["slab_reassign_busy_deletes"] => int(0)
    ["slab_reassign_running"] => int(0)
    ["slabs_moved"] => int(13)
    ["lru_crawler_running"] => int(0)
    ["lru_crawler_starts"] => int(12093)
    ["lru_maintainer_juggles"] => int(3526968)
    ["malloc_fails"] => int(0)
    ["log_worker_dropped"] => int(0)
    ["log_worker_written"] => int(0)
    ["log_watcher_skipped"] => int(0)
    ["log_watcher_sent"] => int(0)
    ["log_watchers"] => int(0)
    ["unexpected_napi_ids"] => int(0)
    ["round_robin_fallback"] => int(0)
    ["bytes"] => int(1853761569)
    ["curr_items"] => int(2237555)
    ["total_items"] => int(3426300)
    ["slab_global_page_pool"] => int(0)
    ["expired_unfetched"] => int(638452)
    ["evicted_unfetched"] => int(0)
    ["evicted_active"] => int(0)
    ["evictions"] => int(0)
    ["reclaimed"] => int(2809)
    ["crawler_reclaimed"] => int(867282)
    ["crawler_items_checked"] => int(457029946)
    ["lrutail_reflocked"] => int(4949844)
    ["moves_to_cold"] => int(11973571)
    ["moves_to_warm"] => int(10394293)
    ["moves_within_lru"] => int(5255254)
    ["direct_reclaims"] => int(0)
    ["lru_bumps_dropped"] => int(0)
  }


  ["cache-1.redacted:11211" ]= array(92) {
    ["pid"] => int(101)
    ["uptime"] => int(1679)
    ["time"] => int(1686604838)
    ["version"] => string(6) "1.6.17"
    ["libevent"] => string(13) "2.1.12-stable"
    ["pointer_size"] => int(64)
    ["rusage_user"] => float(62.687671)
    ["rusage_system"] => float(305.179734)
    ["max_connections"] => int(16384)
    ["curr_connections"] => int(252)
    ["total_connections"] => int(1524)
    ["rejected_connections"] => int(0)
    ["connection_structures"] => int(257)
    ["response_obj_oom"] => int(0)
    ["response_obj_count"] => int(1)
    ["response_obj_bytes"] => int(98304)
    ["read_buf_count"] => int(16880)
    ["read_buf_bytes"] => int(276561920)
    ["read_buf_bytes_free"] => int(276447232)
    ["read_buf_oom"] => int(0)
    ["reserved_fds"] => int(30)
    ["cmd_get"] => int(8520280)
    ["cmd_set"] => int(714165)
    ["cmd_flush"] => int(0)
    ["cmd_touch"] => int(0)
    ["cmd_meta"] => int(0)
    ["get_hits"] => int(6998791)
    ["get_misses"] => int(1521489)
    ["get_expired"] => int(2645)
    ["get_flushed"] => int(0)
    ["delete_misses"] => int(3878)
    ["delete_hits"] => int(3253)
    ["incr_misses"] => int(2)
    ["incr_hits"] => int(32115)
    ["decr_misses"] => int(0)
    ["decr_hits"] => int(0)
    ["cas_misses"] => int(0)
    ["cas_hits"] => int(0)
    ["cas_badval"] => int(0)
    ["touch_hits"] => int(0)
    ["touch_misses"] => int(0)
    ["store_too_large"] => int(1)
    ["store_no_memory"] => int(0)
    ["auth_cmds"] => int(0)
    ["auth_errors"] => int(0)
    ["bytes_read"] => int(1279519896)
    ["bytes_written"] => int(10285267818)
    ["limit_maxbytes"] => int(6442450944)
    ["accepting_conns"] => int(1)
    ["listen_disabled_num"] => int(0)
    ["time_in_listen_disabled_us"] => int(0)
    ["threads"] => int(6)
    ["conn_yields"] => int(0)
    ["hash_power_level"] => int(19)
    ["hash_bytes"] => int(4194304)
    ["hash_is_expanding"] => int(0)
    ["slab_reassign_rescues"] => int(0)
    ["slab_reassign_chunk_rescues"] => int(0)
    ["slab_reassign_evictions_nomem"] => int(0)
    ["slab_reassign_inline_reclaim"] => int(0)
    ["slab_reassign_busy_items"] => int(0)
    ["slab_reassign_busy_deletes"] => int(0)
    ["slab_reassign_running"] => int(0)
    ["slabs_moved"] => int(0)
    ["lru_crawler_running"] => int(0)
    ["lru_crawler_starts"] => int(1233)
    ["lru_maintainer_juggles"] => int(981046)
    ["malloc_fails"] => int(0)
    ["log_worker_dropped"] => int(0)
    ["log_worker_written"] => int(0)
    ["log_watcher_skipped"] => int(0)
    ["log_watcher_sent"] => int(0)
    ["log_watchers"] => int(0)
    ["unexpected_napi_ids"] => int(0)
    ["round_robin_fallback"] => int(0)
    ["bytes"] => int(485391336)
    ["curr_items"] => int(491082)
    ["total_items"] => int(712253)
    ["slab_global_page_pool"] => int(0)
    ["expired_unfetched"] => int(90615)
    ["evicted_unfetched"] => int(0)
    ["evicted_active"] => int(0)
    ["evictions"] => int(0)
    ["reclaimed"] => int(2162)
    ["crawler_reclaimed"] => int(129997)
    ["crawler_items_checked"] => int(28453868)
    ["lrutail_reflocked"] => int(2966)
    ["moves_to_cold"] => int(463056)
    ["moves_to_warm"] => int(92726)
    ["moves_within_lru"] => int(26961)
    ["direct_reclaims"] => int(0)
    ["lru_bumps_dropped"] => int(0)
  }


  ["cache-2.redacted:11211" ]= array(92) {
    ["pid"] => int(66)
    ["uptime"] => int(106459)
    ["time"] => int(1686604838)
    ["version"] => string(6) "1.6.17"
    ["libevent"] => string(13) "2.1.12-stable"
    ["pointer_size"] => int(64)
    ["rusage_user"] => float(6067.660665)
    ["rusage_system"] => float(13887.490941)
    ["max_connections"] => int(16384)
    ["curr_connections"] => int(253)
    ["total_connections"] => int(80870)
    ["rejected_connections"] => int(0)
    ["connection_structures"] => int(367)
    ["response_obj_oom"] => int(0)
    ["response_obj_count"] => int(1)
    ["response_obj_bytes"] => int(98304)
    ["read_buf_count"] => int(60999)
    ["read_buf_bytes"] => int(999407616)
    ["read_buf_bytes_free"] => int(999292928)
    ["read_buf_oom"] => int(0)
    ["reserved_fds"] => int(30)
    ["cmd_get"] => int(548432272)
    ["cmd_set"] => int(25462593)
    ["cmd_flush"] => int(0)
    ["cmd_touch"] => int(0)
    ["cmd_meta"] => int(0)
    ["get_hits"] => int(500285026)
    ["get_misses"] => int(48147246)
    ["get_expired"] => int(100796)
    ["get_flushed"] => int(0)
    ["delete_misses"] => int(391153)
    ["delete_hits"] => int(133534)
    ["incr_misses"] => int(0)
    ["incr_hits"] => int(1243)
    ["decr_misses"] => int(2)
    ["decr_hits"] => int(0)
    ["cas_misses"] => int(0)
    ["cas_hits"] => int(0)
    ["cas_badval"] => int(0)
    ["touch_hits"] => int(0)
    ["touch_misses"] => int(0)
    ["store_too_large"] => int(19)
    ["store_no_memory"] => int(0)
    ["auth_cmds"] => int(0)
    ["auth_errors"] => int(0)
    ["bytes_read"] => int(51165138854)
    ["bytes_written"] => int(550095389278)
    ["limit_maxbytes"] => int(6442450944)
    ["accepting_conns"] => int(1)
    ["listen_disabled_num"] => int(0)
    ["time_in_listen_disabled_us"] => int(0)
    ["threads"] => int(6)
    ["conn_yields"] => int(0)
    ["hash_power_level"] => int(23)
    ["hash_bytes"] => int(67108864)
    ["hash_is_expanding"] => int(0)
    ["slab_reassign_rescues"] => int(118437)
    ["slab_reassign_chunk_rescues"] => int(133)
    ["slab_reassign_evictions_nomem"] => int(0)
    ["slab_reassign_inline_reclaim"] => int(195)
    ["slab_reassign_busy_items"] => int(903)
    ["slab_reassign_busy_deletes"] => int(0)
    ["slab_reassign_running"] => int(0)
    ["slabs_moved"] => int(157)
    ["lru_crawler_running"] => int(0)
    ["lru_crawler_starts"] => int(93533)
    ["lru_maintainer_juggles"] => int(31136001)
    ["malloc_fails"] => int(0)
    ["log_worker_dropped"] => int(0)
    ["log_worker_written"] => int(0)
    ["log_watcher_skipped"] => int(0)
    ["log_watcher_sent"] => int(0)
    ["log_watchers"] => int(0)
    ["unexpected_napi_ids"] => int(0)
    ["round_robin_fallback"] => int(0)
    ["bytes"] => int(5289368519)
    ["curr_items"] => int(10189139)
    ["total_items"] => int(19930526)
    ["slab_global_page_pool"] => int(0)
    ["expired_unfetched"] => int(4987808)
    ["evicted_unfetched"] => int(0)
    ["evicted_active"] => int(0)
    ["evictions"] => int(0)
    ["reclaimed"] => int(8802)
    ["crawler_reclaimed"] => int(6898590)
    ["crawler_items_checked"] => int(4382047003)
    ["lrutail_reflocked"] => int(10950997)
    ["moves_to_cold"] => int(26000887)
    ["moves_to_warm"] => int(17974237)
    ["moves_within_lru"] => int(13197430)
    ["direct_reclaims"] => int(0)
    ["lru_bumps_dropped"] => int(0)
  }


  ["cache-3.redacted:11211" ]= array(92) {
    ["pid"] => int(83)
    ["uptime"] => int(67712)
    ["time"] => int(1686604838)
    ["version"] => string(6) "1.6.17"
    ["libevent"] => string(13) "2.1.12-stable"
    ["pointer_size"] => int(64)
    ["rusage_user"] => float(5111.148219)
    ["rusage_system"] => float(13650.826982)
    ["max_connections"] => int(16384)
    ["curr_connections"] => int(252)
    ["total_connections"] => int(51213)
    ["rejected_connections"] => int(0)
    ["connection_structures"] => int(367)
    ["response_obj_oom"] => int(0)
    ["response_obj_count"] => int(1)
    ["response_obj_bytes"] => int(98304)
    ["read_buf_count"] => int(74652)
    ["read_buf_bytes"] => int(1223098368)
    ["read_buf_bytes_free"] => int(1222983680)
    ["read_buf_oom"] => int(0)
    ["reserved_fds"] => int(30)
    ["cmd_get"] => int(409976287)
    ["cmd_set"] => int(23456907)
    ["cmd_flush"] => int(0)
    ["cmd_touch"] => int(0)
    ["cmd_meta"] => int(0)
    ["get_hits"] => int(375162555)
    ["get_misses"] => int(34813732)
    ["get_expired"] => int(72338)
    ["get_flushed"] => int(0)
    ["delete_misses"] => int(1329547)
    ["delete_hits"] => int(115884)
    ["incr_misses"] => int(0)
    ["incr_hits"] => int(183473)
    ["decr_misses"] => int(4)
    ["decr_hits"] => int(3013)
    ["cas_misses"] => int(0)
    ["cas_hits"] => int(0)
    ["cas_badval"] => int(0)
    ["touch_hits"] => int(0)
    ["touch_misses"] => int(0)
    ["store_too_large"] => int(21)
    ["store_no_memory"] => int(0)
    ["auth_cmds"] => int(0)
    ["auth_errors"] => int(0)
    ["bytes_read"] => int(39891324498)
    ["bytes_written"] => int(411138931240)
    ["limit_maxbytes"] => int(6442450944)
    ["accepting_conns"] => int(1)
    ["listen_disabled_num"] => int(0)
    ["time_in_listen_disabled_us"] => int(0)
    ["threads"] => int(6)
    ["conn_yields"] => int(0)
    ["hash_power_level"] => int(23)
    ["hash_bytes"] => int(67108864)
    ["hash_is_expanding"] => int(0)
    ["slab_reassign_rescues"] => int(69974)
    ["slab_reassign_chunk_rescues"] => int(98)
    ["slab_reassign_evictions_nomem"] => int(0)
    ["slab_reassign_inline_reclaim"] => int(82)
    ["slab_reassign_busy_items"] => int(1)
    ["slab_reassign_busy_deletes"] => int(0)
    ["slab_reassign_running"] => int(0)
    ["slabs_moved"] => int(107)
    ["lru_crawler_running"] => int(0)
    ["lru_crawler_starts"] => int(61190)
    ["lru_maintainer_juggles"] => int(26403500)
    ["malloc_fails"] => int(0)
    ["log_worker_dropped"] => int(0)
    ["log_worker_written"] => int(0)
    ["log_watcher_skipped"] => int(0)
    ["log_watcher_sent"] => int(0)
    ["log_watchers"] => int(0)
    ["unexpected_napi_ids"] => int(0)
    ["round_robin_fallback"] => int(0)
    ["bytes"] => int(5006559690)
    ["curr_items"] => int(8298205)
    ["total_items"] => int(15405514)
    ["slab_global_page_pool"] => int(0)
    ["expired_unfetched"] => int(3507748)
    ["evicted_unfetched"] => int(0)
    ["evicted_active"] => int(0)
    ["evictions"] => int(0)
    ["reclaimed"] => int(6474)
    ["crawler_reclaimed"] => int(4862539)
    ["crawler_items_checked"] => int(3159741123)
    ["lrutail_reflocked"] => int(20407584)
    ["moves_to_cold"] => int(37410901)
    ["moves_to_warm"] => int(31237139)
    ["moves_within_lru"] => int(7253806)
    ["direct_reclaims"] => int(0)
    ["lru_bumps_dropped"] => int(0)
  }
}
  ```
</details>

And then here are the `stats settings`:

<details>
  <summary>Stats settings</summary>
  
  ```
STAT maxbytes 6442450944
STAT maxconns 16384
STAT tcpport 11211
STAT udpport 0
STAT inter NULL
STAT verbosity 0
STAT oldest 0
STAT evictions on
STAT domain_socket NULL
STAT umask 700
STAT shutdown_command no
STAT growth_factor 1.25
STAT chunk_size 48
STAT num_threads 6
STAT num_threads_per_udp 6
STAT stat_key_prefix :
STAT detail_enabled no
STAT reqs_per_event 20
STAT cas_enabled yes
STAT tcp_backlog 1024
STAT binding_protocol auto-negotiate
STAT auth_enabled_sasl no
STAT auth_enabled_ascii no
STAT item_size_max 1048576
STAT maxconns_fast yes
STAT hashpower_init 0
STAT slab_reassign yes
STAT slab_automove 1
STAT slab_automove_ratio 0.80
STAT slab_automove_window 30
STAT slab_chunk_max 524288
STAT lru_crawler yes
STAT lru_crawler_sleep 100
STAT lru_crawler_tocrawl 0
STAT tail_repair_time 0
STAT flush_enabled yes
STAT dump_enabled yes
STAT hash_algorithm murmur3
STAT lru_maintainer_thread yes
STAT lru_segmented yes
STAT hot_lru_pct 20
STAT warm_lru_pct 40
STAT hot_max_factor 0.20
STAT warm_max_factor 2.00
STAT temp_lru no
STAT temporary_ttl 61
STAT idle_timeout 0
STAT watcher_logbuf_size 262144
STAT worker_logbuf_size 65536
STAT read_buf_mem_limit 0
STAT track_sizes no
STAT inline_ascii_response no
STAT ext_item_size 512
STAT ext_item_age 4294967295
STAT ext_low_ttl 0
STAT ext_recache_rate 2000
STAT ext_wbuf_size 4194304
STAT ext_compact_under 0
STAT ext_drop_under 0
STAT ext_max_sleep 1000000
STAT ext_max_frag 0.80
STAT slab_automove_freeratio 0.010
STAT ext_drop_unread no
STAT num_napi_ids (null)
STAT memory_file (null)
  ```
  
</details>

**Comment by WPprodigy** [comment]
Just wrote a script for that :). Can see it increasing here every second:

```
2023-06-12 16:55:03
STAT curr_connections 226
STAT connection_structures 326
STAT read_buf_bytes 276561920
STAT read_buf_bytes_free 276447232
2023-06-12 16:55:04
STAT curr_connections 226
STAT connection_structures 326
STAT read_buf_bytes 345473024
STAT read_buf_bytes_free 301678592
2023-06-12 16:55:05
STAT curr_connections 227
STAT connection_structures 326
STAT read_buf_bytes 378683392
STAT read_buf_bytes_free 329302016
```

**Comment by dormando** [comment]
include the response_obj_* counters?

**Comment by dormando** [comment]
To update the thread, so far it does look like there's an accidental megaget going on.

I need a better way of proving this state to end users and will think about it more; it's pretty obvious to me by combining dozens of counters together and knowing how the code works but it's pretty vague when communicated. I can't add a lot of instrumentation to the `get` path as that is the hottest path in the codebase.

**Comment by WPprodigy** [comment]
Would maybe be slightly less-hot path to keep a counter for the max `response_obj_count` seen?

**Comment by WPprodigy** [comment]
Makes sense!

For anyone that finds this thread in the future, it was indeed some really large multiGets (~50k up to 200k). In summary:

- Really large (unbounded 🙃) DB queries would return say 200k rows of IDs.
- This causes some multiGet lookups for similar volume to get related objects.
- The memcached thread that was lucky enough to get the request would have to significantly allocate some memory for read/response buffers.
- Later on the same type of request would happen again and maybe to a new thread this time, so it would have to allocate the same amount of memory.
- Eventually all threads got the pleasure of serving the request, and the "max" would be reached which explains why `read_buf_bytes` would eventually become static.

For this situation, we'll be implementing client-side controls that batch up multiGets into smaller groups to keep the read buffers from growing out of control. (though of course is also ideal to fix unbounded DB queries and multiGets xD)
